fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/logging/CompletionExceptionLogHelper.java

```diff
@@ -78,9 +99,55 @@ private static Throwable asCauseThrowable(@Nonnull Throwable ex) {
         if (cause == null) {
             return ex;
         }
+
         if (addSuppressed) {
-            cause.addSuppressed(ex);    // Remember the original.
+            synchronized (cause) {
+                Throwable[] suppressedExceptions = cause.getSuppressed();
+                if (suppressedExceptions.length < maxSuppressedCount) {
+                    cause.addSuppressed(ex);
+                } else {
+                    IgnoredSuppressedExceptionCount suppressedCount;
+                    int lastIndex = suppressedExceptions.length - 1;
+                    if (suppressedExceptions[lastIndex] instanceof IgnoredSuppressedExceptionCount) {
+                        suppressedCount = (IgnoredSuppressedExceptionCount) suppressedExceptions[lastIndex];
+                    } else {
+                        suppressedCount = new IgnoredSuppressedExceptionCount();
+                        cause.addSuppressed(suppressedCount);
+                    }
+                    suppressedCount.incrementCount();
+                }
+            }
         }
         return cause;
     }
+
+    /**
+     * A special "exception" to record the number of suppressed exceptions that were not recorded due to the
+     * {@link #maxSuppressedCount}.
+     */
+    public static class IgnoredSuppressedExceptionCount extends Exception {
+        private static final long serialVersionUID = 1L;
+        private static final StackTraceElement[] UNASSIGNED_STACK = new StackTraceElement[0];
+
+        private int count;
+
+        public synchronized void incrementCount() {
+            count++;
+        }
+
+        @Override
+        public StackTraceElement[] getStackTrace() {
+            return UNASSIGNED_STACK;
```

Comment:
It's a little unfortunate that the stack trace here won't be useful, but `fillInStackTrace` will get called in the constructor anyway. I *think* this could actually be addressed by having the default constructor for this class call this constructor: [`Exception(String, Throwable, boolean, boolean)`](https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html#Exception-java.lang.String-java.lang.Throwable-boolean-boolean-) Then if `false` is given to `writableStackTrace`, I think it won't fill in the stack.

Reply:
Great idea! Done.